plugins/file-open-save-new/core/src/main/java/org/pentaho/di/plugins/fileopensave/providers/vfs/VFSFileProvider.java

```diff
@@ -148,81 +198,111 @@ public boolean isAvailable() {
+      throw new FileNotFoundException( fileDomainRoot.getPath(), fileDomainRoot.getProvider() );
     }
 
-    String scheme = vfsConnectionProvider.getProtocol( vfsConnectionDetails );
     for ( VFSRoot root : vfsRoots ) {
       VFSDirectory vfsDirectory = new VFSDirectory();
       vfsDirectory.setName( root.getName() );
       vfsDirectory.setDate( root.getModifiedDate() );
       vfsDirectory.setHasChildren( true );
       vfsDirectory.setCanAddChildren( true );
-      vfsDirectory.setParent( scheme + "://" );
-      vfsDirectory.setDomain( vfsConnectionDetails.getDomain() );
-      vfsDirectory.setConnection( vfsConnectionDetails.getName() );
-      vfsDirectory.setPath( scheme + "://" + root.getName() );
+      vfsDirectory.setParent( fileDomainRoot.getPath() );
+      vfsDirectory.setPath( fileDomainRoot.getPath() + VFSFile.DELIMITER + root.getName() );
       vfsDirectory.setRoot( NAME );
       files.add( vfsDirectory );
     }
-    this.roots.put( file.getConnection(), files );
+    this.roots.put( getKey( fileDomainRoot ), files );
     return files;
   }
 
+  /**
+   * Get key for <code>file</code>
+   * @param vfsFile
+   * @return
+   */
+  protected String getKey( VFSFile vfsFile ) {
+    return vfsFile.getPath();
+  }
+
+  /**
+   * Determines if a <code>connectionDetails</code> has "buckets".
+   * @param connectionDetails
+   * @return
+   */
+  protected boolean hasBuckets( ConnectionDetails connectionDetails ) {
+    return connectionDetails instanceof VFSConnectionDetails && ( (VFSConnectionDetails) connectionDetails).hasBuckets();
+  }
+
   /**
    * @param file
    * @param filters
    * @return
    */
   @Override
   public List<VFSFile> getFiles( VFSFile file, String filters, VariableSpace space ) throws FileException {
-    if ( file.getPath().matches( DOMAIN_ROOT ) ) {
+    if (  isConnectionRoot( file ) ) {
       List<VFSFile> rootFiles = getRoot( file );
       if ( rootFiles != null ) {
         return rootFiles;
       }
     }
     FileObject fileObject;
     try {
-      fileObject = KettleVFS.getInstance( bowl )
-        .getFileObject( file.getPath(), new Variables(), VFSHelper.getOpts( bowl, file.getPath(), file.getConnection(),
-                                                                            space ) );
-    } catch ( KettleFileException | MetaStoreException e ) {
+      fileObject = getFileObject( file, space );
+    } catch ( FileException e ) {
       throw new FileNotFoundException( file.getPath(), TYPE );
     }
     return populateChildren( file, fileObject, filters );
   }
 
+  /**
+   * Determines if <code>file</code>'s path is the URI domain root.
+   * @param file
+   * @return true if <code>file</code>'s path is only at domain, false otherwise.
+   */
+  protected boolean isConnectionRoot( VFSFile file ) {
+    return file != null && !StringUtils.isEmpty( file.getPath() )
+        && file.getPath().matches( CONNECTION_NAME_ROOT_REGEX );
```

Comment:
I think the test would be less reliant on the allowed characters for connection names if we'd just test that the url is `pvfs://one-segment/`. I mean, must have pvfs scheme and a single segment deep. This is how the PUC side code counterpart does it ([see](https://github.com/pentaho/pentaho-scheduler-plugin-ee/blob/07b3e02298c5e605cf487ea33af6fe45c1a23441/core/src/main/java/com/pentaho/platform/genericfile/providers/vfs/connections/BaseVFSConnection.java#L360)).

  Given a URL, this would correspond to having a non-empty authority and an empty path. See `getConnectionName`.

Reply:
Addressed see commit https://github.com/pentaho/pentaho-kettle/commit/3e8bcc9f08d5634a303f5bdcbaf73f80d1811e46